---
title: "Exploratory biomarker analysis"
author: "Malgorzata Nowicka"
date: "`r Sys.Date()`"
output: 
  html_document:
    number_sections: yes
    toc: yes
    toc_depth: 3
    toc_float: 
      collapsed: false
      smooth_scroll: false
vignette: >
  %\VignetteIndexEntry{Exploratory biomarker analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---




```{r setup_knitr, include = TRUE, cache = FALSE}

### Packages to compile rmarkdown
library(knitr)
library(rmarkdown)
### Packages for nice tables
library(kableExtra)


font_size <- 9
cache <- 0


### Set global option for font_size for bkable in this document 

options(bkable_font_size = font_size)
options(bkable_full_width = TRUE)


knitr::opts_chunk$set(cache = cache, cache.comments = FALSE, echo = TRUE, warning = FALSE, message = FALSE, error = FALSE, fig.width = 6, fig.height = 5, fig.align = "center", tidy=TRUE, tidy.opts = list(width.cutoff = 80))


### Necessary for knit_child
knitr::opts_knit$set(output.dir = getwd())


```




```{r library, results="hide", cache = FALSE}


library(ggplot2)

ggplot2::theme_set(cowplot::theme_cowplot())


library(BiomarkerWrappers)


```




```{r library2, echo = FALSE}


devtools::load_all()


```




# Load data


`bdata` object contains random survival and biomarker data. It is not expected to observe any biologically meaningful associations. The goal for this data sets is to solely demonstrate the functionality of this package. 



```{r load_data}


data(bdata)



### Dichotomize gene expression into low and high

genes <- c("BCL2", "GeneA", "GeneB", "GeneC", "GeneD")

data_cat2 <- dplyr::mutate_all(bdata[, genes], wrapper_cut_2groups)
  
colnames(data_cat2) <- paste0(genes, "_cat2")


### Stratify gene expression into quartiles

data_cat4 <- dplyr::mutate_all(bdata[, genes], wrapper_cut_quartiles)
  
colnames(data_cat4) <- paste0(genes, "_cat4")


bdata <- cbind(bdata, data_cat2, data_cat4)








```


## Define variable names



```{r}


variable_names <- format_variable_names(bdata)

variable_names


### Adjust variable names

variable_names[["GeneA_cat2"]] <- "Gene A stratified by median" 



```







# Tile plot





```{r, fig.height=4, fig.width=10, results='asis'}


### Plot generated using geom_tile

wrapper_tile_plot1_core(bdata, y_vars = c("BCL2_by_IHC", "BCL2_by_FISH"))


cat("\n\n ------ \n\n")

### Plot generated using geom_col. It has substantially smaller size when sample size is large.

wrapper_tile_plot2_core(bdata, y_vars = c("BCL2_by_IHC", "BCL2_by_FISH"))




```







# Cox regression



In `wrapper_cox_regression_core_simple()`, covariate_vars Vector with names of covariates that are included in the formula of the simple additive model: `~ covariate_vars[1] + covariate_vars[2] + covariate_vars[3] + ...`.

This function is the core function and should not be used for the analysis, it is a function called by the other functions such as `wrapper_cox_regression_biomarker` and `wrapper_cox_regression_treatment`.

In `wrapper_cox_regression_biomarker` and `wrapper_cox_regression_treatment`, `biomarker_vars` is a vector with biomarker names. For each biomarker a separate Cox regression model is fitted. The results are combined in a single table. 

In `wrapper_cox_regression_biomarker` and `wrapper_cox_regression_treatment`, `adjustment_vars` are added additionally to `biomarker_vars` and `treatment_var` to the model, for example, `~ biomarker_vars[i] + treatment_var + adjustment_vars[1] + adjustment_vars[2] + ...`

`adjustment_vars` would be the official stratification factors from the study plus any other variables you want to use for adjustment, for example, any other known prognostic factors.

`strata_vars`, would be the same as `adjustment_vars`. The difference is that when you define `strata_vars` they are used in Cox regression as "true stratification factors" with the formula `~ biomarker_vars[i] + treatment_var + strata(strata_vars[1] + strata_vars[2] + ...)`

See https://stats.stackexchange.com/questions/256148/stratification-in-cox-model

`strat1_var`, `strat2_var` are used to split the data (I should call them `split1_var`, `split2_var`). Then the analysis is done within the subgroups of each of the variables. For example, if `strat2_var = "PDL1"`, the Cox regression is run separately within the PDL1 low and PDL1 high.




```{r cox_regression, results='asis', fig.width=20, fig.height=4}


### Biomarker effect 

data <- bdata
tte_var <- "PFS"
censor_var <- "PFS_Event"

biomarker_vars <- c("GeneA_cat2", "GeneA")
adjustment_vars <- "IPI"

treatment_var = "Treatment_Arm"


wrapper_cox_regression_biomarker(data, tte_var = tte_var, censor_var = censor_var, biomarker_vars = biomarker_vars, adjustment_vars = adjustment_vars, treatment_var = treatment_var)


## To represent the results with a forest plot

x <- wrapper_cox_regression_biomarker(data, tte_var = tte_var, censor_var = censor_var, biomarker_vars = biomarker_vars, adjustment_vars = adjustment_vars, treatment_var = treatment_var, print_total = FALSE, print_adjpvalues = FALSE)

bforest(x)




### Treatment effect within biomarker subgroups

data <- bdata
tte_var <- "PFS"
censor_var <- "PFS_Event"

treatment_var = "Treatment_Arm"
biomarker_vars <- c("GeneA_cat2", "GeneB_cat2")
adjustment_vars <- "IPI"


wrapper_cox_regression_treatment(data, tte_var = tte_var, censor_var = censor_var, treatment_var = treatment_var, biomarker_vars = biomarker_vars, adjustment_vars = adjustment_vars)
  

## To represent the results with a forest plot

x <- wrapper_cox_regression_treatment(data, tte_var = tte_var, censor_var = censor_var, treatment_var = treatment_var, biomarker_vars = biomarker_vars, adjustment_vars = adjustment_vars, print_total = FALSE, print_adjpvalues = FALSE)


bforest(x)




### Treatment-biomarker interaction effect 


data <- bdata
tte_var <- "PFS"
censor_var <- "PFS_Event"

treatment_var = "Treatment_Arm"
biomarker_vars <- c("GeneA_cat2", "GeneA")
adjustment_vars <- "IPI"



wrapper_cox_regression_interaction(data, tte_var, censor_var, treatment_var = treatment_var, biomarker_vars = biomarker_vars, adjustment_vars = adjustment_vars)
  
  


```






# Log-rank test


```{r log_rank_test, results='asis', fig.width=20, fig.height=5}


### Biomarker effect 

data <- bdata
tte_var <- "PFS"
censor_var <- "PFS_Event"

biomarker_vars <- c("GeneA_cat2", "GeneB_cat2")
treatment_var = "Treatment_Arm"


wrapper_log_rank_test_biomarker(data, tte_var = tte_var, censor_var = censor_var, biomarker_vars = biomarker_vars, treatment_var = treatment_var)


### Treatment effect within biomarker subgroups


data <- bdata
tte_var <- "PFS"
censor_var <- "PFS_Event"

treatment_var = "Treatment_Arm"
biomarker_vars <- c("GeneA_cat2", "GeneB_cat2")


wrapper_log_rank_test_treatment(data, tte_var = tte_var, censor_var = censor_var, treatment_var = treatment_var, biomarker_vars = biomarker_vars)
  


x <- wrapper_log_rank_test_treatment(data, tte_var = tte_var, censor_var = censor_var, treatment_var = treatment_var, biomarker_vars = biomarker_vars, print_nevent = FALSE, print_mst = FALSE, print_total = FALSE)


bkable(x)

bforest(x)




```




# KM plot

## Core

```{r KM_plot}


data <- bdata

tte_var <- "PFS"
censor_var <- "PFS_Event"
covariate_var <- "Treatment_Arm"


wrapper_KM_plot_core(data = data, tte_var = tte_var, censor_var = censor_var, covariate_var = covariate_var)


```


## Core stratified 



```{r KM_plot_strat, fig.width=10, fig.height=15}


data <- bdata
tte_var <- "PFS"
censor_var <- "PFS_Event"
covariate_var <- "GeneA_cat2"

strat1_var = "Treatment_Arm"
strat2_var = "Cell_Of_Origin"
  

wrapper_KM_plot_core_strat(data = data, tte_var = tte_var, censor_var = censor_var, covariate_var = covariate_var, strat1_var = strat1_var, strat2_var = strat2_var, linetypes = c(2, 1))



```


## Display tables with Cox regression results 


```{r}

### Display results from Cox regression as table on the KM plot

data <- bdata

tte_var <- "PFS"
censor_var <- "PFS_Event"
biomarker_var <- "GeneA_cat2"


x <- wrapper_cox_regression_biomarker(data, tte_var = tte_var, censor_var = censor_var, biomarker_vars = biomarker_var, print_total = FALSE, print_adjpvalues = FALSE)


tb <- boutput(x)[, -c(1)]

tb <- tibble::tibble(x = 1, y = 1, label = list(tb))


ggp <- wrapper_KM_plot_core(data = data, tte_var = tte_var, censor_var = censor_var, covariate_var = biomarker_var) 

ggp +
  ggpp::geom_table_npc(data = tb, aes(npcx = x, npcy = y, label = label), size = 3, hjust = 1, vjust = 1, table.theme = ggpp::ttheme_gtminimal)




```



```{r, fig.width=14, fig.height=6}


### Display results from Cox regression as table on the KM plot using one-level stratification

data <- bdata

tte_var <- "PFS"
censor_var <- "PFS_Event"
biomarker_var <- "GeneA_cat2"
treatment_var = "Treatment_Arm"


x <- wrapper_cox_regression_biomarker(data, tte_var = tte_var, censor_var = censor_var, biomarker_vars = biomarker_var, treatment_var = treatment_var, print_total = FALSE, print_adjpvalues = FALSE)

### To have the same level order 

bresults(x)[[treatment_var]] <- factor(bresults(x)[[treatment_var]], levels = levels(data[, treatment_var]))


tb <- boutput(x)[, -c(1, 2)]

tb <- split(tb, bresults(x)[[treatment_var]])

tb <- tibble::tibble(
  x = seq_len(nlevels(data[, treatment_var])) / nlevels(data[, treatment_var]), 
  y = 1, 
  label = tb)


ggp <- wrapper_KM_plot_core_strat(data = data, tte_var = tte_var, censor_var = censor_var, covariate_var = biomarker_var, strat1_var = treatment_var) 

ggp +
  ggpp::geom_table_npc(data = tb, aes(npcx = x, npcy = y, label = label), size = 3, hjust = 1, vjust = 1, table.theme = ggpp::ttheme_gtminimal)



```





```{r, fig.width=14, fig.height=6*3}


### Display results from Cox regression as table on the KM plot using two-level stratification

data <- bdata

tte_var <- "PFS"
censor_var <- "PFS_Event"
biomarker_var <- "GeneA_cat2"
treatment_var = "Treatment_Arm"

strat2_var <- "Cell_Of_Origin"


x <- wrapper_cox_regression_biomarker(data, tte_var = tte_var, censor_var = censor_var, biomarker_vars = biomarker_var, treatment_var = treatment_var, strat2_var = strat2_var, print_total = FALSE, print_adjpvalues = FALSE)

### To have the same level order 

bresults(x)[[treatment_var]] <- factor(bresults(x)[[treatment_var]], levels = levels(data[, treatment_var]))
bresults(x)[[strat2_var]] <- factor(bresults(x)[[strat2_var]], levels = levels(data[, strat2_var]))


tb <- boutput(x)[, -c(1, 2, 3)]

tb <- split(tb, list(bresults(x)[[treatment_var]], bresults(x)[[strat2_var]]))

tb <- tibble::tibble(
  x = rep(seq_len(nlevels(data[, treatment_var])) / nlevels(data[, treatment_var]), times = nlevels(data[, strat2_var])), 
  y = rep(rev(seq_len(nlevels(data[, strat2_var])) / nlevels(data[, strat2_var])), each = nlevels(data[, treatment_var])), 
  label = tb)


ggp <- wrapper_KM_plot_core_strat(data = data, tte_var = tte_var, censor_var = censor_var, covariate_var = biomarker_var, strat1_var = treatment_var, strat2_var = strat2_var, strat1_label_both = FALSE, strat2_label_both = FALSE) 

ggp +
  ggpp::geom_table_npc(data = tb, aes(npcx = x, npcy = y, label = label), size = 3, hjust = 1, vjust = 1, table.theme = ggpp::ttheme_gtminimal)





```




## Curves per treatment arm and biomarker subgroups in one plot


```{r KM_plot_interaction}


data <- bdata
tte_var <- "PFS"
censor_var <- "PFS_Event"

biomarker_var <- "GeneA_cat2"
treatment_var = "Treatment_Arm"


wrapper_KM_plot_interaction(data, tte_var = tte_var, censor_var = censor_var, biomarker_var = biomarker_var, treatment_var = treatment_var)


wrapper_KM_plot_interaction(data, tte_var = tte_var, censor_var = censor_var, biomarker_var = biomarker_var, treatment_var = treatment_var, colors = c("blue", "blue2", "red", "red2"), linetypes = c(2, 1, 2, 1), line_size = 0.5)



```


## Biomarker effect per treatment arm


```{r KM_plot_biomarker, fig.width=10}


data <- bdata
tte_var <- "PFS"
censor_var <- "PFS_Event"

biomarker_var <- "GeneA_cat2"
treatment_var = "Treatment_Arm"


wrapper_KM_plot_biomarker(data, tte_var = tte_var, censor_var = censor_var, biomarker_var = biomarker_var, treatment_var = treatment_var)


```


## Treatment effect per biomarker subgroup


```{r KM_plot_treatment, fig.width=10}


data <- bdata
tte_var <- "PFS"
censor_var <- "PFS_Event"

biomarker_var <- "GeneA_cat2"
treatment_var = "Treatment_Arm"


wrapper_KM_plot_treatment(data, tte_var = tte_var, censor_var = censor_var, biomarker_var = biomarker_var, treatment_var = treatment_var)



```




# Logistic regression


```{r logistic_regression, results='asis'}


### Biomarker effect 

data <- bdata
response_var <- "ORR"

biomarker_vars <- c("GeneA_cat2", "GeneA")
treatment_var = "Treatment_Arm"
adjustment_vars <- "IPI"


wrapper_logistic_regression_biomarker(data, response_var =  response_var, biomarker_vars = biomarker_vars, adjustment_vars = adjustment_vars, treatment_var = treatment_var)




### Treatment effect within biomarker subgroups

data <- bdata
response_var <- "ORR"

treatment_var = "Treatment_Arm"
biomarker_vars <- c("GeneA_cat2", "GeneB_cat2")
adjustment_vars <- "IPI"


wrapper_logistic_regression_treatment(data, response_var =  response_var, treatment_var = treatment_var, biomarker_vars = biomarker_vars, adjustment_vars = adjustment_vars)
  


### Treatment-biomarker interaction effect 


data <- bdata
response_var <- "ORR"

treatment_var = "Treatment_Arm"
biomarker_vars <- c("GeneA_cat2", "GeneA")
adjustment_vars <- "IPI"



wrapper_logistic_regression_interaction(data, response_var =  response_var, treatment_var = treatment_var, biomarker_vars = biomarker_vars, adjustment_vars = adjustment_vars)
  
  


```



# Pearson's test


```{r pearsons_test, results='asis'}


### Biomarker effect 

data <- bdata
response_var <- "ORR"

biomarker_vars <- c("GeneA_cat2")
treatment_var = "Treatment_Arm"

wrapper_pearsons_test_biomarker(data, response_var =  response_var, biomarker_vars = biomarker_vars, treatment_var = treatment_var)



### Treatment effect within biomarker subgroups


data <- bdata
response_var <- "ORR"

biomarker_vars <- c("GeneA_cat2")
treatment_var = "Treatment_Arm"

wrapper_pearsons_test_treatment(data, response_var =  response_var, biomarker_vars = biomarker_vars, treatment_var = treatment_var)





```








# Cochran-Mantel-Haenszel Chi-Squared Test


```{r mantel_test, results='asis'}


### Biomarker effect 

data <- bdata
response_var <- "ORR"

biomarker_vars <- c("GeneA_cat2")
treatment_var = "Treatment_Arm"
strata_vars <- "IPI"


wrapper_pearsons_test_biomarker(data, response_var =  response_var, biomarker_vars = biomarker_vars, treatment_var = treatment_var, strata_vars = strata_vars)



### Treatment effect within biomarker subgroups


data <- bdata
response_var <- "ORR"

biomarker_vars <- c("GeneA_cat2")
treatment_var = "Treatment_Arm"
strata_vars <- "IPI"


wrapper_pearsons_test_treatment(data, response_var =  response_var, biomarker_vars = biomarker_vars, treatment_var = treatment_var, strata_vars = strata_vars)



```




# Barplot response 




```{r, fig.height=5, fig.width=12}


### Biomarker effect 

data <- bdata
response_var <- "ORR"

biomarker_var <- "GeneA_cat2"
treatment_var = "Treatment_Arm"


wrapper_bar_plot_biomarker(data, response_var =  response_var, biomarker_var = biomarker_var, treatment_var = treatment_var, variable_names = variable_names, less_legends = TRUE, ylab = "ORR Proportion (%)", strat1_label_both = FALSE, legend_colors_title = FALSE)



```




```{r, fig.height=5, fig.width=8}


### Treatment effect 

data <- bdata
response_var <- "ORR"

biomarker_var <- "GeneA_cat2"
treatment_var = "Treatment_Arm"


wrapper_bar_plot_treatment(data, response_var =  response_var, biomarker_var = biomarker_var, treatment_var = treatment_var, variable_names = variable_names, less_legends = TRUE, ylab = "R Proportion (%)", strat1_label_both = FALSE, legend_colors_title = FALSE, skip_levels = c("NR"), legend_position = "none")




```






# Barplot


```{r barplot}


data <- bdata

x_var = "GeneA_cat2"
y_var = "Response"


wrapper_bar_plot_core(data = data, x_var = x_var, y_var = y_var)

wrapper_bar_plot_core(data = data, x_var = x_var, y_var = y_var, y_type = "Count")


wrapper_bar_plot_core(data = data, x_var = y_var, y_var = x_var, y_type = "Proportion")

wrapper_bar_plot_core(data = data, x_var = y_var, y_var = x_var, y_type = "Count")



data <- bdata

x_var = "GeneA_cat2"
y_var = "Response"
skip_levels <- c("NE", "PD", "SD")
show_subtotal_proportions <- TRUE

wrapper_bar_plot_core(data = data, x_var = x_var, y_var = y_var, show_subtotal_proportions = show_subtotal_proportions, skip_levels = skip_levels, ylim = c(0, 100))



```



## Stratified 



```{r, fig.height=5, fig.width=10}


data <- bdata

x_var = "GeneA_cat2"
y_var = "Response"
facet_var <- "Treatment_Arm"
  

wrapper_bar_plot_core(data = data, x_var = x_var, y_var = y_var, facet_var = facet_var)


```




```{r, fig.height=15, fig.width=10}


data <- bdata

x_var = "GeneA_cat2"
y_var = "Response"

strat1_var = "Treatment_Arm"
strat2_var = "Cell_Of_Origin"


wrapper_bar_plot_core_strat(data = data, x_var = x_var, y_var = y_var, strat1_var = strat1_var, strat2_var = strat2_var)


wrapper_bar_plot_core_strat(data = data, x_var = x_var, y_var = y_var, strat1_var = strat1_var, strat2_var = strat2_var, less_legends = TRUE)



```




# Complex barplot


```{r Complex_barplot, fig.height=5, fig.width=10}


data <- bdata

x_var <- "Cell_Of_Origin"
y_vars <- c("GeneA_cat2", "GeneB_cat2", "GeneC_cat2")


wrapper_bar_plot_yvars_core_strat(data, x_var = x_var, y_vars = y_vars, values_to = "Gene expression")




data <- bdata

x_var <- "Cell_Of_Origin"
y_vars <- c("GeneA_cat2", "GeneB_cat2", "GeneC_cat2")

skip_levels <- "low"
method <- "dodge"


wrapper_bar_plot_yvars_core_strat(data, x_var = x_var, y_vars = y_vars, skip_levels = skip_levels, method = method, values_to = "Gene expression", names_to = "Gene")





```







# Fisher's test


```{r fishers_test, results='asis'}


data <- bdata

col_var <- "Cell_Of_Origin"
row_vars <- c("GeneA_cat2", "GeneB_cat2", "GeneC_cat2")


wrapper_fishers_test(data, col_var = col_var, row_vars = row_vars)


wrapper_fishers_test(data, col_var = col_var, row_vars = row_vars, margin = 2)




```









# Box plot


```{r Box_plot, fig.height=5, fig.width=7}


data <- bdata

x_var <- "Cell_Of_Origin"
y_var <- "GeneA"


wrapper_box_plot_core(data = data, x_var = x_var, y_var = y_var)




```


## Stratified 

```{r Box_plot_strat, fig.height=5, fig.width=8}



data <- bdata

x_var <- "Cell_Of_Origin"
y_var <- "GeneA"
facet_var <- "Treatment_Arm"
dodge_var = NULL


wrapper_box_plot_core(data = data, x_var = x_var, y_var = y_var, facet_var = facet_var, dodge_var = dodge_var)




data <- bdata

x_var <- "Treatment_Arm"
y_var <- "GeneA"
facet_var <- NULL
dodge_var = "Cell_Of_Origin"


wrapper_box_plot_core(data = data, x_var = x_var, y_var = y_var, facet_var = facet_var, dodge_var = dodge_var)




data <- bdata

x_var <- "Cell_Of_Origin"
y_var <- "GeneA"


strat1_var = "Treatment_Arm"


wrapper_box_plot_core_strat(data = data, x_var = x_var, y_var = y_var, strat1_var = strat1_var)



```





# Complex boxplot




```{r complex_boxplot, fig.width = 10}


data <- bdata
y_vars <- c("GeneA", "GeneB", "GeneC")


wrapper_box_plot_yvars_core_strat(data, y_vars = y_vars)




data <- bdata
y_vars <- c("GeneA", "GeneB", "GeneC")

x_var = "Cell_Of_Origin"
facet_var = NULL
dodge_var = NULL


wrapper_box_plot_yvars_core_strat(data, y_vars = y_vars, x_var = x_var, facet_var = facet_var, dodge_var = dodge_var)



data <- bdata
y_vars <- c("GeneA", "GeneB", "GeneC")

x_var = NULL
facet_var = NULL
dodge_var = "Cell_Of_Origin"


wrapper_box_plot_yvars_core_strat(data, y_vars = y_vars, x_var = x_var, facet_var = facet_var, dodge_var = dodge_var)




```




# Kruskalâ€“Wallis H test


```{r kruskal_test, results='asis'}



data <- bdata

cat_vars <- "Cell_Of_Origin"
num_vars <- c("GeneA", "GeneB")


wrapper_kruskal_test(data, num_vars = num_vars, cat_vars = cat_vars)
  
wrapper_kruskal_test(data, num_vars = num_vars, cat_vars = cat_vars, print_pvalues = FALSE)
  


data <- bdata

cat_vars <- c("Cell_Of_Origin", "IPI", "Treatment_Arm")
num_vars <- "GeneA"


wrapper_kruskal_test(data, num_vars = num_vars, cat_vars = cat_vars)
 
 
  
  

```





# Point plot



```{r Point_plot}


data <- bdata

x_var <- "GeneA"
y_var <- "GeneB"


wrapper_point_plot_core(data = data, x_var = x_var, y_var = y_var)



```


## Stratified


```{r Point_plot_strat, fig.width = 12}


data <- bdata

x_var <- "GeneA"
y_var <- "GeneB"
facet_var <- "Treatment_Arm"
color_point_var <- "Cell_Of_Origin"


wrapper_point_plot_core(data = data, x_var = x_var, y_var = y_var, facet_var = facet_var, color_point_var = color_point_var)




data <- bdata

x_var <- "GeneD"
y_var <- "GeneB"
facet_var <- "Treatment_Arm"
strat1_var <- "Cell_Of_Origin"


wrapper_point_plot_core_strat(data = data, x_var = x_var, y_var = y_var, facet_var = facet_var, strat1_var =  strat1_var)



```





# Characteristics of ITT and BEP


```{r characteristics, results='asis'}



data <- bdata

strat_var <- "Treatment_Arm"
covariate_vars <- c("Age", "IPI")


wrapper_characteristics_core(data, strat_var = strat_var, covariate_vars = covariate_vars)
  
  


bep_vars <- "BEP_RNAseq"
covariate_vars <- c("Age", "IPI")

treatment_var <- "Treatment_Arm"
itt_name = "ITT"


wrapper_characteristics_bep(data, covariate_vars = covariate_vars, bep_vars = bep_vars, treatment_var = treatment_var, itt_name = itt_name)


  
```







```{r}

sessionInfo()

```



















































